### ISSUE TITLE
[Regression 1.8.17] Recursion when computing class usage relations

### ISSUE THREAD

**Issue by nafur** [doxygen/doxygen#7544]
When running doxygen (1.8.17) on our project (https://github.com/smtrat/carl, the doc-pdf target), doxygen stalls at "Computing class usage relations...". It proceeds to accumulate memory and eventually segfaults. It still works with 1.8.16.

The gdb backtrace indicates that we hit an infinite recursion (note the last few stackframes):
```(gdb) bt 20
#0  0x00007ffff7e29440 in std::type_info::operator== (this=0x5555564707e8 <typeinfo for Definition>, __arg=...) at /build/gcc/src/gcc/libstdc++-v3/libsupc++/typeinfo:122
#1  __cxxabiv1::__si_class_type_info::__do_dyncast (this=0x5555564707e8 <typeinfo for Definition>, src2dst=0xffffffffffffffff, access_path=7, dst_type=0x555556470480 <typeinfo for NamespaceDef>, obj_ptr=0x555556f8dc40, src_type=0x5555564707e8 <typeinfo for Definition>, src_ptr=0x555556f8dc40, result=...) at /build/gcc/src/gcc/libstdc++-v3/libsupc++/si_class_type_info.cc:52
#2  0x00007ffff7e29ed0 in __cxxabiv1::__vmi_class_type_info::__do_dyncast (this=0x555556494028 <typeinfo for DefinitionImpl>, src2dst=0xffffffffffffffff, access_path=__cxxabiv1::__class_type_info::__contained_public, dst_type=0x555556470480 <typeinfo for NamespaceDef>, obj_ptr=<optimized out>, src_type=0x5555564707e8 <typeinfo for Definition>, src_ptr=0x555556f8dc40, result=...) at /build/gcc/src/gcc/libstdc++-v3/libsupc++/vmi_class_type_info.cc:160
#3  0x00007ffff7e29ed0 in __cxxabiv1::__vmi_class_type_info::__do_dyncast (this=0x555556486da8 <typeinfo for NamespaceDefImpl>, src2dst=0xffffffffffffffff, access_path=__cxxabiv1::__class_type_info::__contained_public, dst_type=0x555556470480 <typeinfo for NamespaceDef>, obj_ptr=<optimized out>, src_type=0x5555564707e8 <typeinfo for Definition>, src_ptr=0x555556f8dc40, result=...) at /build/gcc/src/gcc/libstdc++-v3/libsupc++/vmi_class_type_info.cc:160
#4  0x00007ffff7e26f7a in __cxxabiv1::__dynamic_cast (src_ptr=0x555556f8dc40, src_type=0x5555564707e8 <typeinfo for Definition>, dst_type=0x555556470480 <typeinfo for NamespaceDef>, src2dst=0xffffffffffffffff) at /build/gcc/src/gcc/libstdc++-v3/libsupc++/dyncast.cc:71
#5  0x00005555558820c4 in followPath(Definition const*, FileDef const*, QCString const&) ()
#6  0x0000555555882eb2 in isAccessibleFromWithExpScope(Definition const*, FileDef const*, Definition const*, QCString const&) ()
#7  0x000055555588351c in isAccessibleFromWithExpScope(Definition const*, FileDef const*, Definition const*, QCString const&) ()
#8  0x000055555588351c in isAccessibleFromWithExpScope(Definition const*, FileDef const*, Definition const*, QCString const&) ()
#9  0x00005555558837f7 in getResolvedSymbol(Definition const*, FileDef const*, Definition*, QCString const&, ArgumentList*, int&, ClassDef const*&, MemberDef const*&, QCString&, QCString&) ()
#10 0x000055555588478d in getResolvedClassRec(Definition const*, FileDef const*, char const*, MemberDef const**, QCString*, QCString*) ()
#11 0x0000555555884d61 in getResolvedClass(Definition const*, FileDef const*, char const*, MemberDef const**, QCString*, bool, bool, QCString*) ()
#12 0x000055555567293a in findClassWithinClassContext(Definition*, ClassDef*, QCString const&) ()
#13 0x0000555555675867 in findClassRelation(Entry const*, Definition*, ClassDef*, BaseInfo*, QDict<int>*, FindBaseClassRelation_Mode, bool) ()
#14 0x000055555567304e in findUsedClassesForClass(Entry const*, Definition*, ClassDef*, ClassDef*, bool, ArgumentList const&, QDict<int>*) ()
#15 0x0000555555674a6a in findTemplateInstanceRelation(Entry const*, Definition*, ClassDef*, QCString const&, QDict<int>*, bool) ()
#16 0x0000555555675a2d in findClassRelation(Entry const*, Definition*, ClassDef*, BaseInfo*, QDict<int>*, FindBaseClassRelation_Mode, bool) ()
#17 0x000055555567304e in findUsedClassesForClass(Entry const*, Definition*, ClassDef*, ClassDef*, bool, ArgumentList const&, QDict<int>*) ()
#18 0x0000555555674a6a in findTemplateInstanceRelation(Entry const*, Definition*, ClassDef*, QCString const&, QDict<int>*, bool) ()
#19 0x0000555555675a2d in findClassRelation(Entry const*, Definition*, ClassDef*, BaseInfo*, QDict<int>*, FindBaseClassRelation_Mode, bool) ()
#20 0x000055555567304e in findUsedClassesForClass(Entry const*, Definition*, ClassDef*, ClassDef*, bool, ArgumentList const&, QDict<int>*) ()
#21 0x0000555555674a6a in findTemplateInstanceRelation(Entry const*, Definition*, ClassDef*, QCString const&, QDict<int>*, bool) ()
#22 0x0000555555675a2d in findClassRelation(Entry const*, Definition*, ClassDef*, BaseInfo*, QDict<int>*, FindBaseClassRelation_Mode, bool) ()
#23 0x000055555567304e in findUsedClassesForClass(Entry const*, Definition*, ClassDef*, ClassDef*, bool, ArgumentList const&, QDict<int>*) ()
#24 0x0000555555674a6a in findTemplateInstanceRelation(Entry const*, Definition*, ClassDef*, QCString const&, QDict<int>*, bool) ()
#25 0x0000555555675a2d in findClassRelation(Entry const*, Definition*, ClassDef*, BaseInfo*, QDict<int>*, FindBaseClassRelation_Mode, bool) ()
#26 0x000055555567304e in findUsedClassesForClass(Entry const*, Definition*, ClassDef*, ClassDef*, bool, ArgumentList const&, QDict<int>*) ()
#27 0x0000555555674a6a in findTemplateInstanceRelation(Entry const*, Definition*, ClassDef*, QCString const&, QDict<int>*, bool) ()
#28 0x0000555555675a2d in findClassRelation(Entry const*, Definition*, ClassDef*, BaseInfo*, QDict<int>*, FindBaseClassRelation_Mode, bool) ()
#29 0x000055555567304e in findUsedClassesForClass(Entry const*, Definition*, ClassDef*, ClassDef*, bool, ArgumentList const&, QDict<int>*) ()
#30 0x0000555555674a6a in findTemplateInstanceRelation(Entry const*, Definition*, ClassDef*, QCString const&, QDict<int>*, bool) ()
#31 0x0000555555675a2d in findClassRelation(Entry const*, Definition*, ClassDef*, BaseInfo*, QDict<int>*, FindBaseClassRelation_Mode, bool) ()
```

I already bisected the commits between Release_1_8_16 and Release_1_8_17 and came up with c38a6fe88ceb3658ae4c09fa9550c770195d9401 being the culprit.

**Comment by doxygen** [comment]
@nafur Please check if the issue has been resolved (but do not close the report yet).

**Comment by nafur** [comment]
Indeed the current HEAD works now, as well as 2faf90d076b3e1ec2d13597341cb8220fe15be9f.
Thanks!

**Comment by doxygen** [comment]
This issue was previously marked 'fixed but not released',
which means it should be fixed in doxygen version 1.8.18.
Please verify if this is indeed the case. Reopen the
issue if you think it is not fixed and please include any additional information
that you think can be relevant (preferably in the form of a self-contained example).
